grafana-plugin/src/state/rootBaseStore.ts

```diff
@@ -192,52 +142,75 @@ export class RootBaseStore {
     this.isUserAnonymous = false;
   }
 
-  async waitForSyncStatus(retryCount = 0) {
-    if (retryCount > SYNC_STATUS_RETRY_LIMIT) {
-      this.retrySync = true;
-      return;
-    }
-
-    getPluginSyncStatus()
-      .then((get_sync_response) => {
-        if (get_sync_response.hasOwnProperty('token_ok')) {
-          this.finishSync(get_sync_response);
-        } else {
-          syncStatusDelay(retryCount + 1).then(() => this.waitForSyncStatus(retryCount + 1));
-        }
-      })
-      .catch((e) => {
-        this.handleSyncException(e);
-      });
-  }
-
-  async setupPlugin(meta: AppPluginMeta<OnCallAppSettings>) {
+  /**
+   * First check to see if the plugin's meta jsonData has an onCallApiUrl saved
+   * If not, tell the user they first need to configure the plugin.
+   *
+   * Otherwise, get the plugin connection status from the OnCall API and check a few pre-conditions:
+   * - plugin must be considered installed by the OnCall API
+   * - user must be not "anonymous" (this is determined by the plugin-proxy)
+   * - the OnCall API must be currently allowing signup
+   * - the user must have an Admin role
+   * If these conditions are all met then trigger a data sync w/ the OnCall backend and poll its response
+   * If the sync is successful, set the error message, otherwise try to load the current user from
+   * the OnCall backend
+   */
+  async setupPlugin(meta: OnCallAppPluginMeta) {
     this.resetStatusToDefault();
+    this.onCallApiUrl = meta.jsonData?.onCallApiUrl;
 
-    if (!meta.jsonData?.onCallApiUrl) {
+    if (!this.onCallApiUrl) {
       this.pluginIsInitialized = false;
       return;
     }
 
-    this.onCallApiUrl = meta.jsonData.onCallApiUrl;
+    const pluginConnectionStatus = await checkIfPluginIsConnected(this.onCallApiUrl);
+    if (typeof pluginConnectionStatus === 'string') {
+      this.initializationError = pluginConnectionStatus;
+      return;
+    }
 
-    let syncStartStatus = await this.startSync();
-    if (syncStartStatus.is_user_anonymous) {
+    const { allow_signup, is_installed, is_user_anonymous } = pluginConnectionStatus;
+    if (is_user_anonymous) {
       this.isUserAnonymous = true;
       return;
-    } else if (!syncStartStatus.is_installed) {
-      if (!syncStartStatus.allow_signup) {
+    } else if (!is_installed) {
+      if (!allow_signup) {
         this.signupAllowedForPlugin = false;
         return;
       }
-      const userRole = await this.getUserRole();
-      if (userRole !== 'Admin') {
+
+      if (!contextSrv.hasRole(OrgRole.Admin)) {
```

Comment:
the `contextSrv` class already provides a utility for this (basically provides access to the currently authenticated grafana user, and some helper methods), which doesn't require making any additional network calls.

See [here](https://github.com/grafana/grafana/blob/main/public/app/core/services/context_srv.ts#L103-L109) for more details.